generator/src/main/java/com/linkedin/pegasus/generator/JavaDataTemplateGenerator.java

```diff
@@ -908,33 +919,87 @@ private void generateWithFieldBodyNested(RecordTemplateSpec.Field field, JDefine
     // For fields with nested types, add a method that allows setting nested mask.
     if (hasNestedFields(field.getSchemaField().getType()))
     {
-      final JMethod withFieldMethod = maskNestedClass.method(JMod.PUBLIC, maskNestedClass, "with" + CodeUtil.capitalize(escapeReserved(field.getSchemaField().getName())));
-      if (!field.getSchemaField().getDoc().isEmpty())
+      final JClass fieldType = generate(field.getType());
+      JClass nestedMaskType = getCodeModel().ref(fieldType.fullName() + "." + PROJECTION_MASK_CLASSNAME);
+      String fieldName = escapeReserved(field.getSchemaField().getName());
+      String maskFieldName = "_" + fieldName + "Mask";
+      JInvocation getDataMap = JExpr.invoke("getDataMap");
+      // Generate a fully typesafe method if the nested type has the new ProjectionMask API.
+      if (hasProjectionMaskApi(fieldType, field.getType()))
+      {
+        JFieldVar maskField = maskNestedClass.fields().get(maskFieldName);
+        if (maskField == null)
+        {
+          maskField = maskNestedClass.field(JMod.PRIVATE, nestedMaskType, maskFieldName);
+        }
+        final JMethod withFieldTypesafeMethod = maskNestedClass.method(JMod.PUBLIC, maskNestedClass, "with" + CodeUtil.capitalize(escapeReserved(
+            fieldName)));
+        if (!field.getSchemaField().getDoc().isEmpty())
+        {
+          withFieldTypesafeMethod.javadoc().append(field.getSchemaField().getDoc());
+        }
+        setDeprecatedAnnotationAndJavadoc(withFieldTypesafeMethod, field.getSchemaField());
+        JVar nestedMask = withFieldTypesafeMethod.param(getCodeModel().ref(Function.class).narrow(nestedMaskType, nestedMaskType), "nestedMask");
+        withFieldTypesafeMethod.body().assign(maskField,
+            nestedMask.invoke("apply").arg(
+                JOp.cond(maskField.eq(JExpr._null()), fieldType.staticInvoke("createMask"), maskField)));
+        withFieldTypesafeMethod.body().invoke(getDataMap, "put")
+            .arg(fieldName)
+            .arg(maskField.invoke("getDataMap"));
+
+        methodBodyCustomizer.accept(withFieldTypesafeMethod);
+        withFieldTypesafeMethod.body()._return(JExpr._this());
+      }
+      // Generate a method that accepts generic mask map if the nested type is from an external source. This is needed
+      // for backwards compatibility.
+      // Note: If the nested type is generated from source PDLs, it will have ProjectionMask.
```

Comment:
nit: this comment can be put on the the method, otherwise you will have to repeat youself across a few places where you call them.